src/cargo/ops/cargo_compile.rs

```diff
@@ -167,27 +187,23 @@ pub fn compile_ws<'a>(ws: &Workspace<'a>,
     }
 
     let profiles = ws.profiles();
-    if spec.len() == 0 {
-        generate_targets(root_package, profiles, mode, filter, release)?;
-    }
 
-    let specs = spec.iter().map(|p| PackageIdSpec::parse(p))
-                                .collect::<CargoResult<Vec<_>>>()?;
-
-    let pair = resolve_dependencies(ws,
+    let resolve = resolve_dependencies(ws,
                                          source,
                                          features,
                                          all_features,
                                          no_default_features,
-                                         &specs)?;
-    let (packages, resolve_with_overrides) = pair;
+                                         &spec)?;
+    let (spec, packages, resolve_with_overrides) = resolve;
 
     let mut pkgids = Vec::new();
     if spec.len() > 0 {
-        for p in spec {
-            pkgids.push(resolve_with_overrides.query(&p)?);
+        for p in spec.iter() {
+            pkgids.push(resolve_with_overrides.query(&p.to_string())?);
         }
     } else {
+        let root_package = ws.current()?;
+        generate_targets(root_package, profiles, mode, filter, release)?;
```

Comment:
I think this was special cased above to get these errors *before* the resolution process happened (which may involve downloads and such). Perhaps that logic could still be special cased?